Use locale_search_str variable in locale.gen_locale and add error handling to integration test

rallytime opened the pull request:
### What does this PR do?
Adds the `locale_search_str` variable to the locale.gen_locale function, rather that having to parse what the `{0}_[1}` string should be in a couple of different places. 

This PR also adds some error checking to the gen_locale execution module integration test. These tests were previously unit tests in the 2016.11 branch and earlier, but were changed to integration tests in the `nitrogen` branch. There are some distributions that don't have charmap files available, which we are logging and returning correctly in the localemod.py file. So some checks have been added to skip the tests when the charmaps are not available.

The assertion that was removed in the `get_locale` test was removed because returning an empty string is a valid return in the function. I added this assertion in #40834, but this caused a new test failure in Fedora 24 where the `get_locale` returns an empty string.

### Tests written?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

Commit:
localemod.gen_locale add locale_search_str var to reduce confusion

Add some spacing for easier readability

Commit:
Skip gen_locale test if charmaps aren't available on the test machine

The error of missing charmaps gets logged to the user. We shouldn't
be running the test for the gen_locale if we don't have charmaps
available on the test machine.

Commit:
Remove assertion against an empty string

Returning an empty string is the expected behavior if the locale
is not set. This happens for example on the Fedora 24 test boxes.

Commit:
Skip test if no charmaps are found

The "loacle -m" command might not return an error, but it might
just be empty. Let's not fail on a test that we're already logging
to the user.

salt-jenkins commented:
@rallytime, thanks for your PR! By analyzing the history of the files in this pull request, we identified @isbm, @anlutro and @jfindlay to be potential reviewers.

isbm commented on salt/modules/localemod.py:
```diff
@@ -315,23 +316,19 @@ def gen_locale(locale, **kwargs):
         )
 
-    if salt.utils.which("locale-gen") is not None:
+    if salt.utils.which('locale-gen') is not None:
```
Not sure if `is not None` is needed, although `salt.utils.which` has an explicit `return None`, which can be actually removed. It either returns a path or nothing.

rallytime replied:
You're right. I'll update that as well.

isbm commented on tests/integration/modules/test_locale.py:
```diff
@@ -27,10 +27,22 @@ class LocaleModuleTest(ModuleCase):
+            self.skipTest('locale charmaps not available. Skipping test.')
+
+        err_msg = 'cannot read character map directory'
+        if char_maps['retcode'] != 0 and err_msg in char_maps['stderr']:
```
Perhaps:
```python
if char_maps['retcode'] and err_msg....
```

rallytime replied:
Yeah, that would be better. I'll update that.

Commit:
Remove "is not None" checks salt.utils.which if statements

These checks are extraneous as salt.utils.which either returns
a path or "None".

Commit:
Clean up integration test checks to be more encompassing

Commit:
Typo: locale should be "locale_info" dictionary
